nbodykit/extensionpoints.py

```diff
@@ -145,78 +88,98 @@ def __init__(cls, name, bases, attrs):
             return
 
         if not hasattr(cls, 'plugin_name'):
-            raise RuntimeError("Plugin class must carry a plugin_name.")
+            raise RuntimeError("Plugin class must carry a 'plugin_name'")
         
         # register, if this plugin isn't yet
-        if cls.plugin_name not in cls.plugins:
-            # add a commandline argument parser for each plugin
-            # NOTE: we don't want every plugin to preparse sys.argv
-            # so set args = ()
-            cls.parser = ArgumentParser(cls.plugin_name, 
-                                        usage=None, 
-                                        add_help=False, 
-                                        formatter_class=cls.help_formatter)
-
-            cls.parser._print_message = _parser_print_message
+        if cls.plugin_name not in cls.registry:
+
+            # initialize the schema and alias it
+            if cls.__init__ == object.__init__:
+                raise ValueError("please define an __init__ method for '%s'" %cls.__name__)
+            cls.__init__.__func__.schema = ConstructorSchema()
```

Comment:
Yes, I'll be adding an update to this PR implementing  sub fields in the
schema as well as a fix for this

On Tuesday, February 23, 2016, Yu Feng <notifications@github.com> wrote:

> In nbodykit/extensionpoints.py
> <https://github.com/bccp/nbodykit/pull/143#discussion_r53882964>:
>
> > -        if cls.plugin_name not in cls.plugins:
> > -            # add a commandline argument parser for each plugin
> > -            # NOTE: we don't want every plugin to preparse sys.argv
> > -            # so set args = ()
> > -            cls.parser = ArgumentParser(cls.plugin_name,
> > -                                        usage=None,
> > -                                        add_help=False,
> > -                                        formatter_class=cls.help_formatter)
> > -
> > -            cls.parser._print_message = _parser_print_message
> > +        if cls.plugin_name not in cls.registry:
> > +
> > +            # initialize the schema and alias it
> > +            if cls.__init__ == object.__init__:
> > +                raise ValueError("please define an __init__ method for '%s'" %cls.__name__)
> > +            cls.__init__.__func__.schema = ConstructorSchema()
>
> Seems like we can remove this __func__ ? It helps python 3 too.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/bccp/nbodykit/pull/143/files#r53882964>.
>